### DISCUSSION TITLE
Couseplay MP support?

### DISCUSSION THREAD

**Discussion by allandata** [Courseplay/courseplay#6973] [Q&A] [answered]
What are the plans for adding some kind of MP support? 

Currently if you run CP on a MP server is more or less useless, all settings are for all farms for all players. There is NO MP thinking in the code as everything is based arround SP gaming. 

We just tried it for 2 days on a MP server and it was caotic on settings and route saving. furthermore the server suddenly became unstable causing sync issues, preventing players from joining server, and leaking memory with many more GC events in the logs, causing the server to suffer under update spikes, with the effects on gameplay that gives.

We all felt it was the best MOD to "enhance" FS19 Helpers, giving lone farmers some help that is capable of not messing up the map. With that in mind we tried it out, and hope for it to work out as we where a quite a few players with good SP experience for the MOD. 

I feel the hunt for new features limits out a whole new way of using the mod on MP servers and i hope in the future that the MP aspect will be incorporated into the MOD.

Thanks for the MOD and hope it will be expanded into the MP aspect of the game.

**Comment by Tensuko** [comment]
> Note that the current version has limited multiplayer support!

Also check the Issues with MP labels.
Most stable MP Version so far is 6.03.00042
https://github.com/Courseplay/courseplay/releases/tag/6.03.00042

We are working on fixing the new MP Errors, but as you noticed, CP was never made to be used in MP and we are getting it working step by step.

The old Coursemanagement Code would need a total rework, wich is a lot of work and will cause new Issues for sure untill it works perfectly again. So for now you have to be a bit more careful on who save and load what course. Example: One big project just made one folder per User and save all course there, make sure other play don't delete or edit your course, communication is the key.

FarmManagement was 1st intruduced in FS19 and CP is way older then FS19 (arround 7-8 years old) and is something, that we won't use, as we don't know what will happen in the next FS.
Settings on the 1st Page are all Vehicle based, while the Settings on the 2nd Page are global.
Mouse for HUD is on both Pages, so you can deactivate it globaly or per Vehicle.
I don't know which setting is problematic to use differently on different farms.

From several players, CP is useable in MP also together with AD.

**Comment by allandata** [comment]
Well then there was all the stability issues that was added to the mix when
we added CP, we suddenly needed daily server restarts, or even multiple,
but as folders are not lockable it's not possible to define where to save a
track, only happens in root, then u have to move it. it's not user friendly
when selveral users are not that familiar with the interface.

Sad to say we lost the fight, but may not even be able to get into the war
before we are on next version of FS. and this is what makes me most sad.

**Reply by Tensuko** [reply]
> but as folders are not lockable it's not possible to define where to save a track, only happens in root, then u have to move it. it's not user friendly when selveral users are not that familiar with the interface.

As I said, it was never made for MP with Farms and stuff.
If you can't trust your players on the server and they don't know what to do, then explain them or not use CP.
Actually, when you use CP+AD, you only need Fieldcourse and that I would always generate new when needed.

**Comment by pvaiko** [comment, marked as answer]
> What are the plans for adding some kind of MP support?

The problem is as usual time, there is currently no dev who could spend time on that. Probably I'm the one who spends most time on CP but I'm not interested in MP. Which also explains the 'hunt' for new features, those are inspired by the problems I see when playing FS in CP myself. 

So yes, we know there's a need but simply put we just have no resources to satisfy those needs.

> We just tried it for 2 days on a MP server and it was caotic on settings and route saving. furthermore the server suddenly became unstable causing sync issues, preventing players from joining server, and leaking memory with many more GC events in the logs, causing the server to suffer under update spikes, with the effects on gameplay that gives.

Do you have evidence for an actual memory leak or you only see the garbage collector eventrs? Not saying leaks are impossible, but course generation and pathfinding need a lot of memory (and CPU, and there's no support for multithreading in the game engine, hence the spikes) and naturally the GC will kick in sooner or later.
